### ISSUE TITLE
SFTP Storages - Unable to access - socket: too many open files 

### ISSUE THREAD

**Issue by mchrisson** [rclone/rclone#4883]
#### What is the problem you are having with rclone?
I use rclone to perform file operations on SFTP storages via my nodeJS based application.

The observations is that,
After rclone process is running for a really long time (I have configured it to run as a service on ubuntu), and there have been many operations performed mostly listing files, etc. I am no longer able to perform any further operations. All I get is the following error (masqueraded).

`{
	"error": "NewFs: couldn't connect SSH: dial tcp 15.97.xxx.xx:22: socket: too many open files",
	"input": {
		"fs": "SFTP-Store-xxxxx:",
		"remote": "xxxx/xxxx/xxxxxx/xxxxxxxxxx"
	},
	"path": "operations/list",
	"status": 500
}`

I found many similar [threads](https://github.com/rclone/rclone/issues/1111), suggesting bumping up the ulimit.
Well I believe that this may just increase the amount of time before which this error starts showing up again, but not solve the issue.

The current ulimit is default to 1024.

Also, here's the output of `lsof | grep "rclone" | wc -l`
8213

Also of `lsof | grep "rclone"` (snippet)
rclone    11323 25901             root 1011u     IPv4          123373227       0t0        TCP \<hostname-of-server-running-rclone>:45468->15.1.xxx.xx:ssh (ESTABLISHED)
rclone    11323 25901             root 1012u     IPv4          123343544       0t0        TCP \<hostname-of-server-running-rclone>:45082->15.1.xxx.xx:ssh (ESTABLISHED)
rclone    11323 25901             root 1013u     IPv4          123313393       0t0        TCP \<hostname-of-server-running-rclone>:47984->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1014u     IPv4          123312025       0t0        TCP \<hostname-of-server-running-rclone>:47986->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1015u     IPv4          123404822       0t0        TCP \<hostname-of-server-running-rclone>:49316->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1016u     IPv4          123343555       0t0        TCP \<hostname-of-server-running-rclone>:48382->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1017u     IPv4          123376162       0t0        TCP \<hostname-of-server-running-rclone>:49032->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1018u     IPv4          123373236       0t0        TCP \<hostname-of-server-running-rclone>:48802->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1019u     IPv4          123373368       0t0        TCP \<hostname-of-server-running-rclone>:49140->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1020u     IPv4          123373238       0t0        TCP \<hostname-of-server-running-rclone>:48818->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1021u     IPv4          123344101       0t0        TCP \<hostname-of-server-running-rclone>:48370->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1022u     IPv4          123344102       0t0        TCP \<hostname-of-server-running-rclone>:48372->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)
rclone    11323 25901             root 1023u     IPv4          123343550       0t0        TCP \<hostname-of-server-running-rclone>:48374->\<hostname-of-SFTP-server>:ssh (ESTABLISHED)

I see too many connections that stay "ESTABLISHED". I wonder why these don't close and rather stay connected.

The output of `lsof | grep "rclone" | grep -i "close"` is empty.

Perhaps, there is a file descriptor leak OR may be SSH connections being kept active after every operation, thus growing in number unnecessarily.

Ofcourse, restarting rclone resolves the issue, but I don't intent to be restarting it.

#### What is your rclone version (output from `rclone version`)
rclone v1.53.2
- os/arch: linux/amd64
- go version: go1.15.3

#### Which OS you are using and how many bits (e.g. Windows 7, 64 bit)
Ubuntu 18.04.3 LTS, 64 bit

####  Which cloud storage system are you using? (e.g. Google Drive)
SFTP

#### The command you were trying to run (e.g. `rclone copy /tmp remote:tmp`)
Using rclone API to list files or any other operations on the remote storage for that matter.

#### A log from the command with the `-vv` flag (e.g. output from `rclone -vv copy /tmp remote:tmp`)
Below is the response from 'operations/list' on storage: "SFTP-Store-xxxxx" using API.

`{
	"error": "NewFs: couldn't connect SSH: dial tcp 15.97.xxx.xx:22: socket: too many open files",
	"input": {
		"fs": "SFTP-Store-xxxxx:",
		"remote": "xxxx/xxxx/xxxxxx/xxxxxxxxxx"
	},
	"path": "operations/list",
	"status": 500
}`

**Comment by ivandeex** [comment]
This was reported in #2614 but we failed to reproduce it.

Let's give it another try. Please contract the open file limit for test sake, run your commands and send API requests repeatedly

```
ulimit -n 100
rclone rcd your_sftp:path -vvv --log-file bug.log
[open 2nd terminal]
pgrep -a -f rclone
curl http://localhost:5572 ???
[repeat API requests]
```

After it has run for a while and accumulated some open files, kill rclone with `kill -QUIT RclonePID`. 
This will produce a backtrace which is likely very long
Please attach your commands, log and trace after redacting secrets here (or on gist.hithub.com or pastebin).

Thank you.

**Comment by mchrisson** [comment]
I set the ulimit to 100 and tried but was finding it difficult to see the error show up in logs.
So I set the ulimit to **50**. Following which I ran:
`/usr/bin/rclone rcd --rc-user=service-user --rc-pass=password --rc-web-gui --rc-addr=0.0.0.0:5572 -vvv --log-file /home/ubuntu/logs/bug.log`

You can find the log trace [here](https://gist.github.com/mchrisson/d2e595705119f1b8b8d098ddf1b74fe2)

Also max count I got from running `lsof | grep "rclone" | wc -l` was `240`.

I did see this `socket: too many open files` in many places in the log immediately. But then it went away.

Also note, the multiple SFTP storages I created all point to the same remote machine. But different directories. _(this is how it fits my use-case)_
also, I had a scheduler triggering multiple request to my nodeJS API which in turn talks to rclone API to 'operation/list' or move files.

**Comment by ncw** [comment]
Thanks for the log - very useful.

From the traceback I see groups of 26 goroutines which are likely the problem. Ultimately these are all causeed by `sftp.(*Fs).sftpConnection` which is the sftp backends routine to create a new connection.

I think each connection must be using 2 sockets to get to your ulimit of 50.

The SFTP backend takes care to re-use connections in its connection pool but it doesn't ever purge connections.

```
26: IO wait [0~13 minutes] [Created by ssh.newClientTransport @ handshake.go:134]
    poll    netpoll.go:220        runtime_pollWait(*, 0x72, #6)
    poll    fd_poll_runtime.go:87 (*pollDesc).wait(*, 0x72, *, 0x1000, 0x1000)
    poll    fd_poll_runtime.go:92 (*pollDesc).waitRead(...)
    poll    fd_unix.go:159        (*FD).Read(*, *, 0x1000, 0x1000, 0, 0, 0)
    net     fd_posix.go:55        (*netFD).Read(*, *, 0x1000, 0x1000, 0x5d243e, *, 0x10)
    net     net.go:182            (*conn).Read(*, *, 0x1000, 0x1000, 0, 0, 0)
    bufio   bufio.go:227          (*Reader).Read(*, *, 0x4, 0x4, *, *, *)
    io      io.go:314             ReadAtLeast(#4, *, *, 0x4, 0x4, 0x4, 0x40d950, *, *)
    io      io.go:333             ReadFull(...)
    ssh     cipher.go:373         (*gcmCipher).readCipherPacket(*, *, #4, *, *, *, *, 0, 0)
    ssh     transport.go:130      (*connectionState).readPacket(*, *, *, 0x405d2e, *, *, *)
    ssh     transport.go:114      (*transport).readPacket(*, *, *, *, 0, 0)
    ssh     handshake.go:374      (*handshakeTransport).readOnePacket(*, *, *, *, *, 0, 0)
    ssh     handshake.go:197      (*handshakeTransport).readLoop(*)
26: sync.Cond.Wait [0~13 minutes] [Created by sftp.NewClientPipe @ client.go:173]
    runtime proc.go:312           goparkunlock(...)
    sync    sema.go:513           runtime_notifyListWait(*, *)
    sync    cond.go:56            (*Cond).Wait(*)
    ssh     buffer.go:94          (*buffer).Read(*, *, 0x4, 0x4, 0, 0, 0)
    ssh     channel.go:351        (*channel).ReadExtended(*, *, 0x4, 0x4, 0, *, 0, *)
    ssh     channel.go:528        (*channel).Read(*, *, 0x4, 0x4, *, 0x405edc, *)
    io      io.go:314             ReadAtLeast(#5, *, *, 0x4, 0x4, 0x4, 0x405835, *, *)
    io      io.go:333             ReadFull(...)
    sftp    packet.go:144         recvPacket(#5, *, *, *, 0x1, *, 0, 0)
    sftp    conn.go:22            (*conn).recvPacket(...)
    sftp    conn.go:75            (*clientConn).recv(*, 0, 0)
    sftp    conn.go:60            (*clientConn).loop(*)
26: sync.Cond.Wait [13~14 minutes] [Created by ssh.NewClient @ client.go:61]
    runtime proc.go:312           goparkunlock(...)
    sync    sema.go:513           runtime_notifyListWait(*, *)
    sync    cond.go:56            (*Cond).Wait(*)
    ssh     mux.go:110            (*mux).Wait(*, 0, 0)
    ssh     client.go:62          NewClient.func1(*)
26: sync.Cond.Wait [13 minutes] [Created by sftp.(*Fs).sftpConnection @ sftp.go:312]
    runtime proc.go:312           goparkunlock(...)
    sync    sema.go:513           runtime_notifyListWait(*, *)
    sync    cond.go:56            (*Cond).Wait(*)
    ssh     mux.go:110            (*mux).Wait(*, 0, 0)
    sftp    sftp.go:274           (*conn).wait(*)
26: chan receive [0~13 minutes] [Created by ssh.newMux @ mux.go:128]
    ssh     handshake.go:187      (*handshakeTransport).readPacket(*, *, *, *, 0, 0)
    ssh     mux.go:215            (*mux).onePacket(*, 0, 0)
    ssh     mux.go:190            (*mux).loop(*)
26: chan receive [13 minutes] [Created by ssh.newSession @ session.go:592]
    ssh     session.go:423        (*Session).wait(*, *, *, *)
    ssh     session.go:593        newSession.func1(*, *)
26: chan receive [13 minutes] [Created by ssh.NewClient @ client.go:59]
    ssh     client.go:138         (*Client).handleGlobalRequests(*, *)
26: select [13~14 minutes] [Created by ssh.newClientTransport @ handshake.go:135]
    ssh     handshake.go:268      (*handshakeTransport).kexLoop(*)
26: chan receive [13~14 minutes] [Created by ssh.NewClient @ client.go:60]
    ssh     client.go:147         (*Client).handleChannelOpens(*, *)
```

So there could be 3 causes of this I can think of

1. The SFTP backend was used concurrently by 26 routines at some point - this would have required 26 connections.
2. There is an SFTP connection leak somewhere, so connections are being made but not returned to the pool.
3. You are creating lots of different SFTP backends - currently the connection pool is never being cleared

I think looking at your log it is option 3 - you are continually creating SFTP backends. It look like from the log that the backends aren't being re-used though - each request is creating a new backend which is wrong.

It looks like you are creating lots of backends very quickly - I wonder if there is a race condition there...

Looking at the code, this looks like the cause of the problem.

Try this. It should make sure you only create one backend each time. However this still won't flush the connection pool when the backend is removed from the cache - that should be done too.

[v1.54.0-beta.4992.9b9bf0568.fix-4883-cache](https://beta.rclone.org/branch/fix-4883-cache/v1.54.0-beta.4992.9b9bf0568.fix-4883-cache/) on branch [fix-4883-cache](https://github.com/rclone/rclone/tree/fix-4883-cache) (uploaded in 15-30 mins)

**Comment by OIITCONZ** [comment]
@ncw  You wrote: Try this. It should make sure you only create one backend each time. However this still won't flush the connection pool when the backend is removed from the cache - that should be done too. 
v1.54.0-beta.4992.9b9bf0568.fix-4883-cache on branch fix-4883-cache (uploaded in 15-30 mins)

I've uploaded that file and run it (Windows Server 2012 using sync to Asustor NAS via sftp)  I am still getting the issue
My max limit is over 7600 and i am hitting that many threads / processes in under three minutes. this then crashes the Asustor - requiring reboot. 

C:\Utils\rclone>rclone --version
 > rclone v1.54.0-beta.4992.9b9bf0568.fix-4883-cache
 > - os/arch: windows/386
 > - go version: go1.15.6
- 
Unfortuntely that files doesnt fix the issue
. Stopping the rclone process short of the asustor crqashing does not release the threads / processes    10 minutes in and I've still got numerous processes.  Running pkill sorts them but then the rclone throws a wobbly and cant restart. 

HTH
Shane

**Comment by ncw** [comment]
@OIITCONZ thanks for testing. Can you run with `-vv` and attach the log please? I'll analyse and see where it is going wrong now! Thanks.

**Comment by mchrisson** [comment]
@ncw , just some more information.

I kept a watch on the count from these two commands:
`netstat -anlp | grep rclone | wc -l`
and
`lsof | grep "rclone" | wc -l`

This just keeps growing.
When the count of netstat reaches 1017, that's when we start to see the "too many open files" error popping up. meanwhile the lsof count is around 7000-9000.

I believe that on each request like "operations/list" a new connection is established rather than re-using existing.

Also another interesting thing,
Even after I removed a remote-config (SFTP type)
Running netstat I could still see a SSH connection ESTABLISHED to that remote system and this connection never got released.

**Comment by ncw** [comment]
@mchrisson can you post a log with -vv from the test version I posted above?

**Comment by ncw** [comment]
@mchrisson just wondering if you saw the message above?

**Comment by mchrisson** [comment]
Hi @ncw, Took me time.

I have tried, v1.54.0-beta.4992.9b9bf0568.fix-4883-cache

But with no luck. The problem persists. The list of open files reached 1028. That's when it starts to give the errors.

```
root@server:/home/ubuntu# lsof -c rclone | wc -l
1028
root@server:/home/ubuntu# netstat -anlp | grep rclone | wc -l
1018
```

I ran rc operation/list :
```
root@server:/home/ubuntu/rclone# rclone rc operations/list --json '{"fs": "89664090-8552-478b-a4dd-28a189660c1a:", "remote":"PQ_Dropbox_mount"}' --rc-user=service-user --rc-pass=password -vvv --log-file /home/ubuntu/rclone/bug.log
root@server:/home/ubuntu/rclone# cat bug.log
2021/02/03 16:34:08 DEBUG : rclone: Version "v1.54.0-beta.4992.9b9bf0568.fix-4883-cache" starting with parameters ["rclone" "rc" "operations/list" "--json" "{\"fs\": \"89664090-8552-478b-a4dd-28a189660c1a:\", \"remote\":\"PQ_Dropbox_mount\"}" "--user=service-user" "--pass=password" "-vvv" "--log-file" "/home/ubuntu/rclone/bug.log"]
2021/02/03 16:34:19 DEBUG : 4 go routines active
2021/02/03 16:34:19 Failed to rc: Failed to read rc response: 500 Internal Server Error: {
        "error": "NewFs: couldn't connect SSH: dial tcp 15.97.199.24:22: socket: too many open files",
        "input": {
                "fs": "89664090-8552-478b-a4dd-28a189660c1a:",
                "remote": "PQ_Dropbox_mount"
        },
        "path": "operations/list",
        "status": 500
}
```

This took 2 days to start giving errors. But I sadly missed to include the -vvv and log while running the rcd.

Gathering the logs from rcd. Would post them soon.

**Comment by mchrisson** [comment]
Hi @ncw,

[here](https://gist.github.com/mchrisson/2dbb85027285a28b9953aa430405c538) is the log.

I am not able to get the backtrace after killing rclone. Re-trying again. Would post it here.

**Comment by mchrisson** [comment]
Hi @ncw ,

[here](https://gist.github.com/mchrisson/4a4d2670989bf26b639de2a416114e80) is the log with the back trace.

I had lowered the ulimit to 50. Took me few hours. Ofcourse if left at default of 1024. Then it takes a couple of days to start seeing the errors.

**Steps to reproduce:**
I had 9 SFTP remotes added.
At an interval of roughly 15-20 mins I made rc operation of "operation/ls" to list the directories on each remote.
Repeated this couple of time till I started see the error.
```
{"data":{"error":"NewFs: couldn't connect SSH: dial tcp 15.97.xxx.xx:22: socket: too many open files","input":{"fs":"e46a986b-ee3d-4da6-8327-383f517aaa8b:","remote":"PQ_Dropbox_mount/folder1/folder2/folderXX12"},"path":"operations/list","status":500}}
```

**Comment by ncw** [comment]
There are only 8 SFTP backends in use there which is good.

A quick look at the traceback shows that there are 38 SFTP connections active.

Am I right in thinking the ulimit for that log is 50?

I made this beta which extends the cache lifetime to 24 hours.  My theory is that the SFTP backends keep getting created and not re-used. So if this works then the theory is correct

[v1.55.0-beta.5154.544757a33.fix-4883-cache](https://beta.rclone.org/branch/fix-4883-cache/v1.55.0-beta.5154.544757a33.fix-4883-cache/) on branch [fix-4883-cache](https://github.com/rclone/rclone/tree/fix-4883-cache) (uploaded in 15-30 mins)

**Comment by mchrisson** [comment]
@ncw. Yes. The ulimit was at 50.

Would try out this beta and post the logs

**Comment by mchrisson** [comment]
@ncw,

This seems to be good.

[here](https://gist.github.com/mchrisson/261e27adb9f3f8949f6c0f524e5bb217) is the log.

But in the above log the ulimit were the default.

One question thought,
What would happen after 24 hours? Would new SFTP backends get created?
In other words, Is the cache lifetime reset back to 24hrs on a reuse ?

**Comment by ncw** [comment]
Thanks for testing.

> Note, On one occasion I had seen active SFTP(SSH) connections to a server, that was added as a SFTP remote but then deleted from the config (via rc). Yet the SSH connections were active to that server even days later. perhaps something to consider and look at.

Yes this is a much better way of fixing the problem - namely getting the sftp backends to close idle connections.

This beta closes idle connections after 1 minute. You can configure that with `--sftp-idle-timeout` or the `idle_timeout` config flag.

I think that should fix your problem. Can you give it a try please?

Thanks

[v1.55.0-beta.5175.e43e03161.fix-4883-sftp-idle-timeout](https://beta.rclone.org/branch/fix-4883-sftp-idle-timeout/v1.55.0-beta.5175.e43e03161.fix-4883-sftp-idle-timeout/) on branch [fix-4883-sftp-idle-timeout](https://github.com/rclone/rclone/tree/fix-4883-sftp-idle-timeout) (uploaded in 15-30 mins)

If this technique works, then I should put it into the other backends with connection pools - I think the FTP backend is the only other one.

I will come back to the backend cache during the v1.55 dev cycle, but I don't think it is the best solution for you.

**Comment by mchrisson** [comment]
Hi @ncw,

This is working well.

[here](https://gist.github.com/mchrisson/e419dbcbf19e721678214fb1a51da1ae) is the log.

On idle the the count from `lsof -c rclone | wc -l` was 15 and on use it reached 23, which is in line with the number of SFTP remotes I have (8).

This kind off works for me, as mainly the number of connections is not simply growing eating up the open files limit.
For my use-case, I have a scheduler that does an "operation/ls" on every remote at an schedule of 1hour. So its fine if idle connections are closed. As well on the other hand if I lower the schedule to may be 5-10 mins then I can accordingly adjust the --sftp-idle-timeout to fit my need.

I will continue with some more tests / scenarios. And let know, if anything else pops up.

Thanks

**Comment by ncw** [comment]
Great - thank you for testing.

I've merged this to master now which means it will be in [the latest beta](https://beta.rclone.org) in 15-30 mins and released in v1.55

I also did the same for the FTP backend.

Let me know if you find any further problems - thanks
